### DISCUSSION TITLE
How to handle muti middlewares?

### DISCUSSION THREAD

**Discussion by PengBoUESTC** [vercel/next.js#51111] [Help]
### Summary

if I have different middleware for different route, how to realize this ?

**Comment by devjiwonchoi** [comment]
It is **not possible** to set different middleware per route since the Middleware file should be located at the root segment of your project.

> Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

Also, there is no reason to set different middleware since the purpose of the middleware itself is to manage the overall request of your project.

> Middleware will be invoked for every route in your project.

Therefore, you can locate your `middleware.js | ts` on the root segment of your project which is the **same level** your `/app` or `/page` folder is located, and manage the routes as follows:

> For example, at the same level as pages or app, or inside src if applicable.

```js
// middleware.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
 
export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    return NextResponse.rewrite(new URL('/about-2', request.url))
  }
 
  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    return NextResponse.rewrite(new URL('/dashboard/user', request.url))
  }
}
```

You can also use the [Matcher](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher) in order to catch the routes you want to run the middleware.

> matcher allows you to filter Middleware to run on specific paths.

```js

// middlware.ts

export const config = {
  matcher: ['/about/:path*', '/dashboard/:path*'], // Matcher regex
}

// ...rest of your middleware code
```

For more `Matcher` regex, see [here](https://github.com/pillarjs/path-to-regexp#path-to-regexp-1).

## Conclusion

You do not need `middleware` files per route since Next.js allows `middleware` to be located only at the root segment of your project.
You can configure it by the incoming pathname of the `request` or use the `Matcher`.

## Reference
- [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware)
- [Matcher](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher)
- [Matcher Path to Regex](https://github.com/pillarjs/path-to-regexp#path-to-regexp-1)
